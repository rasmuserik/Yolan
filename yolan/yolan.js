var tokenRegEx = RegExp.call(RegExp, "\\s*(\\[|\\]|(\\\\.|[^\\s\\[\\]])+)", "g");

var unescapeRegEx = RegExp.call(RegExp, "\\\\(.)", "g");

var escapeRegEx = RegExp.call(RegExp, "[\\\\\\[\\] \\n\\r]", "g");

var yolan = {};

yolan.tokenize = function(str) {
    var result = [];
    str.replace(tokenRegEx, function(_, token) {
        return result.push(token);
    });
    return result;
};

yolan.parse = function(tokens) {
    tokens = tokens.reverse();
    var stack = [];
    var current = [ "do" ];
    while (tokens["length"]) {
        var token = tokens.pop();
        if (token === "[") {
            stack.push(current);
            current = [];
        } else if (token === "]") {
            var t = current;
            current = stack.pop();
            current.push(t);
        } else if (true) {
            current.push(token.replace(unescapeRegEx, function(_, a) {
                return a;
            }));
        }
    }
    return current;
};

yolan.nspace = function(n) {
    var result = [];
    while (0 < n) {
        n = n - 1;
        result.push(" ");
    }
    return result.join("");
};

var indent = 0;

var screenWidth = 78;

var indentStep = 2;

yolan.prettyprint = function(ast) {
    if (typeof ast === "string") {
        return ast.replace(escapeRegEx, function(s) {
            return "\\" + s;
        });
    }
    if (0 === ast["length"]) {
        return "[]";
    }
    indent = indent + indentStep;
    var pos = indent;
    strs = ast.map(yolan["prettyprint"]);
    if (pos + strs.join()["length"] + 1 < screenWidth) {
        indent = indent - indentStep;
        return "[" + strs.join(" ") + "]";
    }
    var space = "\n" + yolan.nspace(indent);
    var result = [];
    result.push("[");
    result.push(strs[0]);
    pos = pos + strs[0]["length"] + 1;
    var i = 1;
    var forceNewLine = false;
    var currentIsString = true;
    while (i < ast["length"]) {
        var prevIsString = currentIsString;
        currentIsString = typeof ast[i] === "string";
        forceNewLine = false;
        if (!prevIsString || !currentIsString) {
            forceNewLine = true;
        }
        if (i < 2) {
            forceNewLine = false;
        }
        var prevIsString = currentIsString;
        result.push(" ");
        if (forceNewLine || screenWidth < pos + strs[i]["length"]) {
            result.pop();
            result.push(space);
            pos = indent;
        }
        result.push(strs[i]);
        pos = pos + strs[i]["length"] + 1;
        i = i + 1;
    }
    result.push("]");
    indent = indent - indentStep;
    return result.join("");
};

var compileJS = {
    JsTypeOf: function(syn, syn1) {
        return "typeof " + yolan.toJS(syn1);
    },
    "do": function(syn) {
        return syn.slice(1).map(yolan["toJS"]).join(";");
    },
    def: function(syn, syn1) {
        return "var " + syn1 + "=" + yolan.toJS(syn[2]);
    },
    set: function(syn, syn1) {
        return syn1 + "=" + yolan.toJS(syn[2]);
    },
    object: function(syn, syn1) {
        return "{" + syn.slice(1).map(function(pair) {
            return JSON.stringify(pair[0].toString()) + ":" + yolan.toJS(pair[1]);
        }).join(",") + "}";
    },
    array: function(syn, syn1) {
        return "[" + syn.slice(1).map(yolan["toJS"]).join(",") + "]";
    },
    fn: function(syn, syn1) {
        return "function(" + syn1.join(",") + "){" + syn.slice(2, -1).map(yolan["toJS"]).join(";") + ";return " + yolan.toJS(syn[syn["length"] - 1]) + "}";
    },
    "while": function(syn, syn1) {
        return "while(" + yolan.toJS(syn1) + "){" + syn.slice(2).map(yolan["toJS"]).join(";") + "}";
    },
    cond: function(syn) {
        return syn.slice(1).map(function(pair) {
            return "if(" + yolan.toJS(pair[0]) + "){" + pair.slice(1).map(yolan["toJS"]).join(";") + "}";
        }).join("else ");
    },
    "if": function(syn, syn1) {
        return "if(" + yolan.toJS(syn1) + "){" + syn.slice(2).map(yolan["toJS"]).join(";") + "}";
    },
    "#": function() {
        return "";
    },
    "return": function(syn, syn1) {
        return "return " + yolan.toJS(syn1);
    },
    "throw": function(syn, syn1) {
        return "throw " + yolan.toJS(syn1);
    },
    "+": function(syn) {
        return syn.slice(1).map(yolan["toJS"]).join("+");
    },
    "-": function(syn) {
        return syn.slice(1).map(yolan["toJS"]).join("-");
    },
    and: function(syn) {
        return syn.slice(1).map(yolan["toJS"]).join("&&");
    },
    or: function(syn) {
        return syn.slice(1).map(yolan["toJS"]).join("||");
    },
    not: function(syn, syn1) {
        return "!" + yolan.toJS(syn1);
    },
    "<": function(syn, syn1) {
        return yolan.toJS(syn1) + "<" + yolan.toJS(syn[2]);
    },
    "<=": function(syn, syn1) {
        return yolan.toJS(syn1) + "<=" + yolan.toJS(syn[2]);
    },
    "eq?": function(syn, syn1) {
        return yolan.toJS(syn1) + "===" + yolan.toJS(syn[2]);
    }
};

yolan.toJS = function(syn) {
    var syn0 = syn[0];
    var syn1 = syn[1];
    if (typeof syn === "string") {
        return syn;
    } else if (compileJS[syn0]) {
        return compileJS[syn0].call(null, syn, syn1);
    } else if (syn1 === "set") {
        return yolan.toJS(syn0) + "." + syn[2] + "=" + yolan.toJS(syn[3]);
    } else if (syn1 === "get") {
        return yolan.toJS(syn0) + "[" + yolan.toJS(syn[2]) + "]";
    } else if (typeof syn0 === "string") {
        return yolan.toJS(syn0) + "." + syn1 + "(" + syn.slice(2).map(yolan["toJS"]).join(",") + ")";
    }
    return yolan.toJS(syn0) + "." + syn1 + "(" + syn.slice(2).map(yolan["toJS"]).join(",") + ")";
};

var fs = require.call(null, "fs");

var action = process["argv"][2];

if (action === "compile") {
    fs.readFile(process["argv"][3], "utf8", function(err, data) {
        if (err) {
            return err;
        }
        var js = yolan.toJS(yolan.parse(yolan.tokenize(data)));
        var uglify = require.call(null, "uglify-js");
        var jsp = uglify["parser"];
        var pro = uglify["uglify"];
        var ast = jsp.parse(js);
        js = pro.gen_code(ast, {
            beautify: true
        });
        return fs.writeFile(process["argv"][4], js, function(err, data) {
            if (err) {
                return err;
            }
            return true;
        });
    });
}

if (action === "prettyprint") {
    fs.readFile(process["argv"][3], "utf8", function(err, data) {
        if (err) {
            return err;
        }
        var ast = yolan.parse(yolan.tokenize(data));
        var src = ast.slice(1).map(yolan["prettyprint"]).join("\n\n");
        return fs.writeFile(process["argv"][4], src, function(err, data) {
            if (err) {
                return err;
            }
            return true;
        });
    });
}