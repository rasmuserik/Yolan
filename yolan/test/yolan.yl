[# bug: double escaping needed - both for js-strings and yolan strings ]
[def tokenRegEx [RegExp call RegExp '\\\\s*(\\\\\[|\\\\\]|(\\\\\\\\.|\[^\\\\s\\\\\[\\\\\]\])+)' 'g']]

[def unescapeRegEx [RegExp call RegExp '\\\\\\\\(.)' 'g']]

[def yolan [object]]

[yolan set tokenize [fn [str]
    [def result [array]]
    [str replace tokenRegEx [fn [_ token]
        [result push token]]]
    result]]

[yolan set parse [fn [tokens]
    [set tokens [tokens reverse]]
    [def stack [array]]
    [def current [array 'do']]
    [while [tokens get 'length']
        [def token [tokens pop]]
        [cond
            [[eq? token '\[']
                [stack push current]
                [set current [array]]]
            [[eq? token '\]']
                [def t current]
                [set current [stack pop]]
                [current push t]]
            [true [current push [token replace unescapeRegEx [fn [_ a] a]]]]]]
    current]]

[def indent 0]
[yolan set nspace [fn [n]
    [def result [array]]
    [while [< 0 n] [set n [- n 1]] [result push '\ ']]
    [result join '']]]
    
[yolan set prettyprint [fn [ast]
    [def pos 0]
    [def width 78]
    [def indentStep 2]
    [def acc [array]]
    [def pp [fn [ast]
        [cond 
            [[eq? 'string' [JsTypeOf ast]] [acc push ast ]]
            [[Array isArray ast] 
                [def str [+ '\[' [[ast map yolan.prettyprint] join '\ ' ] '\]']]
                [set indent [+ indent 1]]
                [cond 
                  [[< [- width indent [str get 'length']] 0]
                    [acc push '\[']
                    [acc push [yolan prettyprint [ast get 0]]]
                    [acc push '\ ']
                    [acc push [[[ast slice 1] map yolan.prettyprint] join [+ '\\n' [yolan nspace indent]]]]
                    [acc push '\]']]
                  [true [acc push str]]]
                [set indent [- indent 1]]
            ]
        ]
        true
    ]]
    [pp call null ast]
    [acc join '']
]]

[def compileJS [object
    [JsTypeOf [fn [syn syn1] [+ 'typeof\ ' syn1]]]
    [do [fn [syn] [[[syn slice 1] map [yolan get 'toJS']] join ';']]]
    [def [fn [syn syn1] [+ 'var\ ' syn1 '=' [yolan toJS [syn get 2]]]]]
    [set [fn [syn syn1] [+ syn1 '=' [yolan toJS [syn get 2]]]]]
    [object [fn [syn syn1] [+ '{' 
        [[[syn slice 1] map [fn [pair] 
            [+ [JSON stringify [[pair get 0] toString]] ':' [yolan toJS [pair get 1]]]]]
        join ','] '}']]]
    [array [fn [syn syn1] [+ '\[' [[[syn slice 1] map [yolan get 'toJS']] join ',']'\]']]]
    [fn [fn [syn syn1] [+ 'function(' [syn1 join ','] '){' [[[syn slice 2 -1] map [yolan get 'toJS']] join ';'] ';return\ ' [yolan toJS [syn get [- [syn get 'length'] 1]]] '}']]]
    [while [fn [syn syn1] [+ 'while(' [yolan toJS syn1] '){' [[[syn slice 2] map [yolan get 'toJS']] join ';'] '}']]]
    [cond [fn [syn] 
        [[[syn slice 1] map 
            [fn [pair] [+
                'if(' 
                [yolan toJS [pair get 0]] 
                '){' 
                [[[pair slice 1] map [yolan get 'toJS']] join ';'] 
                '}']]] join 'else\ ']]]
     [# [fn [] '']]
     [return [fn [syn syn1] [+ 'return\ ' [yolan toJS syn1]]]]
     [throw [fn [syn syn1] [+ 'throw\ ' [yolan toJS syn1]]]]
     [+ [fn [syn] [[[syn slice 1] map [yolan get 'toJS']] join '+']]]
     [- [fn [syn] [[[syn slice 1] map [yolan get 'toJS']] join '-']]]
     [and [fn [syn] [[[syn slice 1] map [yolan get 'toJS']] join '&&']]]
     [or [fn [syn] [[[syn slice 1] map [yolan get 'toJS']] join '||']]]
     [not [fn [syn] [+ '!' [yolan toJS syn1]]]]
     [< [fn [syn syn1] [+ [yolan toJS syn1] '<' [yolan toJS [syn get 2]]]]]
     [<= [fn [syn syn1] [+ [yolan toJS syn1] '<=' [yolan toJS [syn get 2]]]]]
     [eq? [fn [syn syn1] [+ [yolan toJS syn1] '===' [yolan toJS [syn get 2]]]]]
]]

[yolan set toJS [fn [syn]
    [def syn0 [syn get 0]]
    [def syn1 [syn get 1]]
    [cond
        [[eq? [JsTypeOf syn] 'string'] [return syn]]
        [[compileJS get syn0] [return [[compileJS get syn0] call null syn syn1]]]
        [[eq? syn1 'set'] [return [+ [yolan toJS syn0] '.' [syn get 2] '=' [yolan toJS [syn get 3]]]]]
        [[eq? syn1 'get'] [return [+ [yolan toJS syn0] '\[' [yolan toJS [syn get 2]] '\]']]]
        [[eq? [JsTypeOf syn0] 'string'] [ return [+ [yolan toJS syn0] '.' syn1 '(' [[[syn slice 2] map [yolan get 'toJS']] join ','] ')']]]
    ]
    [+ [yolan toJS syn0] '.' syn1 '(' [[[syn slice 2] map [yolan get 'toJS']] join ','] ')']
]]

[def fs [require call null 'fs']]

[def action [[process get 'argv'] get 2]]

[cond 
    [[eq? action 'compile']
        [fs readFile [[process get 'argv'] get 3] 'utf8' [fn [err data]
            [cond [err [return err]]]
            [# compile yolan to js]
            [def js [yolan toJS [yolan parse [yolan tokenize data]]]]
            [# reformat js]
            [def uglify [require call null 'uglify-js']]
            [def jsp [uglify get 'parser']]
            [def pro [uglify get 'uglify']]
            [def ast [jsp parse js]]
            [set js [pro gen_code ast [object [beautify true]]]]
            [fs writeFile [[process get 'argv'] get 4] js [fn [err data] [cond [err [return err]]] true]]]]]
    [[eq? action 'prettyprint']
        [fs readFile [[process get 'argv'] get 3] 'utf8' [fn [err data]
            [cond [err [return err]]]
            [def ast [yolan parse [yolan tokenize data]]]
            [console log [yolan prettyprint ast]]
            true
        ]]
    ]
]
