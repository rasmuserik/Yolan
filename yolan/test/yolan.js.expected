var tokenRegEx = RegExp.call(RegExp, "\\s*(\\[|\\]|(\\\\.|[^\\s\\[\\]])+)", "g");

var unescapeRegEx = RegExp.call(RegExp, "\\\\(.)", "g");

var escapeRegEx = RegExp.call(RegExp, "[\\\\\\[\\] \\n\\r]", "g");

var yolan = {};

yolan.tokenize = function(str) {
    var result = [];
    str.replace(tokenRegEx, function(_, token) {
        return result.push(token);
    });
    return result;
};

yolan.parse = function(tokens) {
    tokens = tokens.reverse();
    var stack = [];
    var current = [ "do" ];
    while (tokens["length"]) {
        var token = tokens.pop();
        if (token === "[") {
            stack.push(current);
            current = [];
        } else if (token === "]") {
            var t = current;
            current = stack.pop();
            current.push(t);
        } else if (true) {
            current.push(token.replace(unescapeRegEx, function(_, a) {
                return a;
            }));
        }
    }
    return current;
};

var indent = 0;

yolan.nspace = function(n) {
    var result = [];
    while (0 < n) {
        n = n - 1;
        result.push(" ");
    }
    return result.join("");
};

yolan.prettyprint = function(ast) {
    var pos = 0;
    var width = 78;
    var indentStep = 2;
    var acc = [];
    var pp = function(ast) {
        if ("string" === typeof ast) {
            acc.push(ast.replace(escapeRegEx, function(s) {
                return "\\" + s;
            }));
        } else if (Array.isArray(ast)) {
            var str = "[" + ast.map(yolan["prettyprint"]).join(" ") + "]";
            indent = indent + 1;
            if (width - indent - str["length"] < 0) {
                acc.push("[");
                acc.push(yolan.prettyprint(ast[0]));
                acc.push(" ");
                acc.push(ast.slice(1).map(yolan["prettyprint"]).join("\n" + yolan.nspace(indent)));
                acc.push("]");
            } else if (true) {
                acc.push(str);
            }
            indent = indent - 1;
        }
        return true;
    };
    pp.call(null, ast);
    return acc.join("");
};

var compileJS = {
    JsTypeOf: function(syn, syn1) {
        return "typeof " + syn1;
    },
    "do": function(syn) {
        return syn.slice(1).map(yolan["toJS"]).join(";");
    },
    def: function(syn, syn1) {
        return "var " + syn1 + "=" + yolan.toJS(syn[2]);
    },
    set: function(syn, syn1) {
        return syn1 + "=" + yolan.toJS(syn[2]);
    },
    object: function(syn, syn1) {
        return "{" + syn.slice(1).map(function(pair) {
            return JSON.stringify(pair[0].toString()) + ":" + yolan.toJS(pair[1]);
        }).join(",") + "}";
    },
    array: function(syn, syn1) {
        return "[" + syn.slice(1).map(yolan["toJS"]).join(",") + "]";
    },
    fn: function(syn, syn1) {
        return "function(" + syn1.join(",") + "){" + syn.slice(2, -1).map(yolan["toJS"]).join(";") + ";return " + yolan.toJS(syn[syn["length"] - 1]) + "}";
    },
    "while": function(syn, syn1) {
        return "while(" + yolan.toJS(syn1) + "){" + syn.slice(2).map(yolan["toJS"]).join(";") + "}";
    },
    cond: function(syn) {
        return syn.slice(1).map(function(pair) {
            return "if(" + yolan.toJS(pair[0]) + "){" + pair.slice(1).map(yolan["toJS"]).join(";") + "}";
        }).join("else ");
    },
    "#": function() {
        return "";
    },
    "return": function(syn, syn1) {
        return "return " + yolan.toJS(syn1);
    },
    "throw": function(syn, syn1) {
        return "throw " + yolan.toJS(syn1);
    },
    "+": function(syn) {
        return syn.slice(1).map(yolan["toJS"]).join("+");
    },
    "-": function(syn) {
        return syn.slice(1).map(yolan["toJS"]).join("-");
    },
    and: function(syn) {
        return syn.slice(1).map(yolan["toJS"]).join("&&");
    },
    or: function(syn) {
        return syn.slice(1).map(yolan["toJS"]).join("||");
    },
    not: function(syn) {
        return "!" + yolan.toJS(syn1);
    },
    "<": function(syn, syn1) {
        return yolan.toJS(syn1) + "<" + yolan.toJS(syn[2]);
    },
    "<=": function(syn, syn1) {
        return yolan.toJS(syn1) + "<=" + yolan.toJS(syn[2]);
    },
    "eq?": function(syn, syn1) {
        return yolan.toJS(syn1) + "===" + yolan.toJS(syn[2]);
    }
};

yolan.toJS = function(syn) {
    var syn0 = syn[0];
    var syn1 = syn[1];
    if (typeof syn === "string") {
        return syn;
    } else if (compileJS[syn0]) {
        return compileJS[syn0].call(null, syn, syn1);
    } else if (syn1 === "set") {
        return yolan.toJS(syn0) + "." + syn[2] + "=" + yolan.toJS(syn[3]);
    } else if (syn1 === "get") {
        return yolan.toJS(syn0) + "[" + yolan.toJS(syn[2]) + "]";
    } else if (typeof syn0 === "string") {
        return yolan.toJS(syn0) + "." + syn1 + "(" + syn.slice(2).map(yolan["toJS"]).join(",") + ")";
    }
    return yolan.toJS(syn0) + "." + syn1 + "(" + syn.slice(2).map(yolan["toJS"]).join(",") + ")";
};

var fs = require.call(null, "fs");

var action = process["argv"][2];

if (action === "compile") {
    fs.readFile(process["argv"][3], "utf8", function(err, data) {
        if (err) {
            return err;
        }
        var js = yolan.toJS(yolan.parse(yolan.tokenize(data)));
        var uglify = require.call(null, "uglify-js");
        var jsp = uglify["parser"];
        var pro = uglify["uglify"];
        var ast = jsp.parse(js);
        js = pro.gen_code(ast, {
            beautify: true
        });
        return fs.writeFile(process["argv"][4], js, function(err, data) {
            if (err) {
                return err;
            }
            return true;
        });
    });
} else if (action === "prettyprint") {
    fs.readFile(process["argv"][3], "utf8", function(err, data) {
        if (err) {
            return err;
        }
        var ast = yolan.parse(yolan.tokenize(data));
        var src = ast.slice(1).map(yolan["prettyprint"]).join("\n\n");
        fs.writeFile(process["argv"][4], src, function(err, data) {
            if (err) {
                return err;
            }
            return true;
        });
        console.log(src);
        return true;
    });
}